File: html/en/7gui/timer.c
#include <stdlib.h>
#include <iup.h>

int btn_reset_cb( Ihandle *self )
{
  Ihandle* text = IupGetDialogChild(self, "GAUGE");
  Ihandle* val = IupGetDialogChild(self, "VAL");
  Ihandle* dial = IupGetDialog(self);
  Ihandle* timer = (char *)IupGetAttribute(dial, "TIMER");

  IupSetAttribute(timer, "RUN", "NO");
  IupSetAttribute(timer, "RUN", "YES");

  IupSetAttribute(timer, "TOTALELAPSEDTIME", "0");
  IupSetAttribute(val, "VALUE", "30");

  return IUP_DEFAULT;
}

int val_valuechanged_cb(Ihandle *self)
{
  Ihandle* gauge = IupGetDialogChild(self, "GAUGE");
  Ihandle* dial = IupGetDialog(self);
  Ihandle* timer = (char *) IupGetAttribute(dial, "TIMER");
  double value = IupGetDouble(self, "VALUE");
  double totalTime = IupGetDouble(timer, "TOTALELAPSEDTIME");

  IupSetDouble(gauge, "MAX", value);

  if (totalTime/1000. < value)
    IupSetAttribute(timer, "RUN", "YES");
  else
    IupSetAttribute(gauge, "VALUE", IupGetAttribute(gauge, "MAX"));

  return IUP_DEFAULT;
}

int timer_cb(Ihandle *self)
{
  Ihandle* gauge = (Ihandle *)IupGetAttribute(self, "GAUGE");
  Ihandle* val = (Ihandle *)IupGetAttribute(self, "VAL");
  Ihandle* label = (Ihandle *)IupGetAttribute(self, "LABEL");
  double elapsedTime = IupGetDouble(self, "ELAPSEDTIME");
  double totaltime = IupGetDouble(self, "TOTALELAPSEDTIME");

  if ((totaltime + elapsedTime) / 1000. >= IupGetDouble(val, "VALUE"))
  {
    IupSetAttribute(self, "RUN", "NO");
    IupSetStrf(self, "TOTALELAPSEDTIME", "%.2lf", totaltime + elapsedTime);
    IupSetStrf(label, "TITLE", "%.2fs", IupGetDouble(val, "VALUE"));
    IupSetStrf(gauge, "VALUE", "%.2fs", IupGetDouble(val, "VALUE"));
  }
  else
  {
    IupSetStrf(label, "TITLE", "%.2fs", (totaltime + elapsedTime) / 1000.);
    IupSetStrf(gauge, "VALUE", "%.2fs", (totaltime + elapsedTime) / 1000.);
  }

  return IUP_DEFAULT;
}

int main(int argc, char **argv)
{
  Ihandle *dlg, *gauge, *label, *glabel, *val, *vlable, *button, *vbox, *hbox1, *hbox2, *timer;

  IupOpen(&argc, &argv);

  timer = IupTimer();
  IupSetAttribute(timer, "TIME", "100");
  IupSetAttribute(timer, "TOTALELAPSEDTIME", "0");
  IupSetCallback(timer, "ACTION_CB", (Icallback)timer_cb);

  glabel = IupLabel("Elapsed Time:");
  gauge = IupGauge();
  IupSetAttribute(gauge, "NAME", "GAUGE");
  IupSetAttribute(gauge, "MIN", ".1");
  IupSetAttribute(gauge, "MAX", "30");
  IupSetAttribute(gauge, "VALUE", "30");
  IupSetAttribute(gauge, "EXPAND", "HORIZONTAL");
  IupSetAttribute(gauge, "SHOWTEXT", "NO");
  label = IupLabel("0.0s");
  IupSetAttribute(label, "NAME", "LABEL");
  IupSetAttribute(label, "EXPAND", "HORIZONTAL");
  vlable = IupLabel("Duration:");
  val = IupVal("HORIZONTAL");
  IupSetAttribute(val, "EXPAND", "HORIZONTAL");
  IupSetAttribute(val, "NAME", "VAL");
  IupSetAttribute(val, "MAX", "60.0");
  IupSetAttribute(val, "STEP", ".1");
  IupSetAttribute(val, "VALUE", "30.0");
  IupSetAttribute(val, "PAGESTEP", ".5");
  button = IupButton("Reset", NULL);
  IupSetAttribute(button, "EXPAND", "HORIZONTAL");
  IupSetAttribute(button, "NAME", "BUTTON");
  hbox1 = IupHbox(glabel, gauge, NULL);
  IupSetAttribute(hbox1, "MARGIN", "0x0");
  IupSetAttribute(hbox1, "ALIGNMENT", "ACENTER");
  hbox2 = IupHbox(vlable, val, NULL);
  IupSetAttribute(hbox2, "MARGIN", "0x0");
  IupSetAttribute(hbox2, "ALIGNMENT", "ACENTER");
  vbox = IupVbox(
    hbox1,
    label,
    hbox2,
    button,
    NULL);
  IupSetAttribute(vbox, "MARGIN", "10x10");
  IupSetAttribute(vbox, "GAP", "10");
  dlg = IupDialog(vbox);
  IupSetAttribute(dlg, "TITLE", "Timer");
  IupSetAttribute(dlg, "SIZE", "200");
  IupSetAttribute(dlg, "TIMER", (char *) timer);

  IupSetAttribute(timer, "GAUGE", (char *)gauge);
  IupSetAttribute(timer, "LABEL", (char *)label);
  IupSetAttribute(timer, "VAL", (char *)val);

  /* Registers callbacks */
  IupSetCallback(val, "VALUECHANGED_CB", (Icallback)val_valuechanged_cb);
  IupSetCallback(button, "ACTION", (Icallback)btn_reset_cb);

  IupShowXY(dlg, IUP_CENTER, IUP_CENTER);

  IupMainLoop();

  IupClose();
  return EXIT_SUCCESS;
}

File: html/en/7gui/timer.lua
require("iuplua")

--********************************** Main *****************************************

timer = iup.timer{time = 100, run = "YES", totalelapsedtime = 0}
glabel = iup.label{title = "Elapsed Time:"}
gauge = iup.gauge{min = .1, max = 30, expand = "HORIZONATAL", showtext="NO"}
label = iup.label{text = "0.0s", expand = "HORIZONTAL"}
vlable = iup.label{title = "Duration:"}
val = iup.val{"HORIZONTAL"; expand = "HORIZONTAL", min = "0", max = "60", step = ".1", value = "30.0", pagestep = ".5"}
button = iup.button{title = "Reset", expand = "HORIZONTAL"}

hbox1 = iup.hbox{glabel, gauge; margin = "0x0", alignment = "ACENTER"}
hbox2 = iup.hbox{vlable, val; margin = "0x0", alignment = "ACENTER"}
vbox = iup.vbox{hbox1, label, hbox2, button; margin = "10x10", gap = 10}

dlg = iup.dialog{vbox, title = "Timer", size = "200"}

function timer:action_cb()

  local elapsedTime = self.elapsedtime or 0
  local totaltime = self.totalelapsedtime
  
  local t = (totaltime + elapsedTime) / 1000.
  local valValue = val.value
  
  if t >= tonumber(valValue) then
    self.run = "NO"
    self.totalelapsedtime =  totaltime + elapsedTime
    label.title = string.format("%.2fs", val.value)
    gauge.value = val.value
  else
    label.title = string.format("%.2fs", (totaltime + elapsedTime) / 1000.)
    gauge.value = (totaltime + elapsedTime) / 1000.
	print(gauge.value.." - "..type(gauge.value))
  end
end

function val:valuechanged_cb()

  local totalTime = timer.totalelapsedtime

  gauge.max = self.value

  if totalTime/1000. < tonumber(self.value) then
    timer.run = "YES"
  else
    gauge.value = gauge.max
  end
end

function button:action()

  timer.run = "NO"

  timer.totalelapsedtime = 0
  val.value = "30."
  gauge.value = 0
  gauge.min = 0
  gauge.max = 60

  timer.run = "YES"
end

dlg:showxy( iup.CENTER, iup.CENTER )

if (iup.MainLoopLevel()==0) then
  iup.MainLoop()
end
